stdlib/array.ml

```diff
@@ -344,3 +344,69 @@ let of_rev_list = function
 let of_seq i =
   let l = Seq.fold_left (fun acc x -> x::acc) [] i in
   of_rev_list l
+
+(* Array.filter, Float.Array.filter, Bytes.filter are almost the same code:
+   if one needs to be modified, the other will probably need
+   it too *)
+
+let filter p arg =
+  let size_arg = length arg in
+  if size_arg = 0 then [| |] else
+    (* create the bit vector *)
+    let bv = Bytes.create ((size_arg - 1) lsr 3 + 1) in
+    (* count elements e such that p(e) *)
+    let size_res = ref 0 in
+    (* fill bv, one byte at a time *)
+    for i = 0 to (size_arg lsr 3) - 1 do
+      let i8 = i lsl 3 in
+      (* p(arg.(8*i + k)) is (bv.[i] land (1 lsl k)) <> 0 *)
+      let c =
+        0 lor
+        (if p (unsafe_get arg i8)         then   1 else 0) lor
+        (if p (unsafe_get arg (i8 lor 1)) then   2 else 0) lor
+        (if p (unsafe_get arg (i8 lor 2)) then   4 else 0) lor
+        (if p (unsafe_get arg (i8 lor 3)) then   8 else 0) lor
+        (if p (unsafe_get arg (i8 lor 4)) then  16 else 0) lor
+        (if p (unsafe_get arg (i8 lor 5)) then  32 else 0) lor
+        (if p (unsafe_get arg (i8 lor 6)) then  64 else 0) lor
+        (if p (unsafe_get arg (i8 lor 7)) then 128 else 0) in
+      Bytes.unsafe_set bv i (Char.unsafe_chr c);
```

Comment:
I think these functions could be made faster by eliminating the branchiness here, along these lines:

```ocaml
          (Bool.to_int (p (unsafe_get arg (i8 lor 0))) lsl 0) lor
          (Bool.to_int (p (unsafe_get arg (i8 lor 1))) lsl 1) lor
          (Bool.to_int (p (unsafe_get arg (i8 lor 2))) lsl 2) lor
          ...
```

if the implementation of `Bool.to_int` were also changed to:

```ocaml
external to_int : bool -> int = "%identity"
```

Reply:
Also the evaluation order being unspecified, we don't know which `p` will be called first.
I would expect that kind of function (like other higher-order functions from Array/Bytes/...) to evaluate `p` on all elements in array order.

Reply:
Switching to the following code:

```ocaml
      let c0 = Bool.to_int2 (p (unsafe_get arg (i8 lor 0))) lsl 0 in
      let c1 = Bool.to_int2 (p (unsafe_get arg (i8 lor 1))) lsl 1 in
      let c2 = Bool.to_int2 (p (unsafe_get arg (i8 lor 2))) lsl 2 in
      let c3 = Bool.to_int2 (p (unsafe_get arg (i8 lor 3))) lsl 3 in
      let c4 = Bool.to_int2 (p (unsafe_get arg (i8 lor 4))) lsl 4 in
      let c5 = Bool.to_int2 (p (unsafe_get arg (i8 lor 5))) lsl 5 in
      let c6 = Bool.to_int2 (p (unsafe_get arg (i8 lor 6))) lsl 6 in
      let c7 = Bool.to_int2 (p (unsafe_get arg (i8 lor 7))) lsl 7 in
      let c = c0 lor c1 lor c2 lor c3 lor c4 lor c5 lor c6 lor c7 in
```

results in a 30% speed-up on my machine for moderately sized (10,000-1,000,000 element) byte sequences with a predicate that selects about half the elements.

Reply:
What about:
```ocaml
      let c = Bool.to_int (p (unsafe_get arg (i8 lor 0))) lsl 0 in
      let c = (Bool.to_int (p (unsafe_get arg (i8 lor 1))) lsl 1) lor c in
      let c = (Bool.to_int (p (unsafe_get arg (i8 lor 2))) lsl 2) lor c in
      let c = (Bool.to_int (p (unsafe_get arg (i8 lor 3))) lsl 3) lor c in
      let c = (Bool.to_int (p (unsafe_get arg (i8 lor 4))) lsl 4) lor c in
      let c = (Bool.to_int (p (unsafe_get arg (i8 lor 5))) lsl 5) lor c in
      let c = (Bool.to_int (p (unsafe_get arg (i8 lor 6))) lsl 6) lor c in
      let c = (Bool.to_int (p (unsafe_get arg (i8 lor 7))) lsl 7) lor c in
```
This should put less pressure on the register allocator as values are short-lived, there are less value to save, and the dependency created seem negligible in face of the memory load and the call.

Reply:
That seems approximately the same speed as the original version.

Reply:
What about a version that also accumulates directly the number of elements (i.e. the sum of the 8 Bool.to_int values) instead of calling popcount?

Reply:
Accumulating the number of elements directly seems to make things slightly slower than the version I posted above.

A further speed improvement is available from unrolling the second loop (i.e. the loop that copies the elements across).  My version above takes around 70% of the running time of the original (i.e. it's about 1.4x the speed).
After unrolling the second loop the running time is around 63% of the original (i.e. it's about 1.6x the speed).

Reply:
I have one more version that's around 1.8x the speed of the original (55% of the running time) by rolling the second loop rather aggressively extent:

```ocaml
        begin match c land 0xf with
        | 0b0000 -> ()
        | 0b0001 -> unsafe_set result (!i_res+0) (unsafe_get arg (base lor 0));
                    i_res := !i_res + 1
        | 0b0010 -> unsafe_set result (!i_res+0) (unsafe_get arg (base lor 1));
                    i_res := !i_res + 1
        | 0b0011 -> unsafe_set result (!i_res+0) (unsafe_get arg (base lor 0));
                    unsafe_set result (!i_res+1) (unsafe_get arg (base lor 1));
                    i_res := !i_res + 2
         ...
         begin match (c lsr 4) land 0xf with
         ...
```

(For this last one I think the speedup is probably not sufficient to justify the amount of additional code.)

Reply:
My feeling is that we should stop at the version you posted above ("Switching to the following code:...    let c0 = Bool.to_int (p (unsafe_get arg (i8 lor 0))) lsl 0 in ... results in a 30% speed-up"). Do you agree?